Downstream eval

lllAlexanderlll opened the pull request:
* New conversion endpoint
* Create HF wrappers for our GPT2 Model and Config
* Create test for new conversion endpoint 
* Fix broken test from main

mrudat-iais commented:
@le1nux we just adjusted our HF checkpoint conversion such that it does not use FSDP anymore. to do so, we added a new class called PytorchToDiscCheckpointing. feel free to have a look at the PR!

lllAlexanderlll commented on tests/data/checkpoint/checkpoint_for_testing/model_config.yaml:
```diff
@@ -0,0 +1,144 @@
+      dataset:
+        type_hint: PackedMemMapDatasetContinuous
+        config: 
+          raw_data_path: /home/richard-rutmann/s3/projects/remote_projects/modalities/data/sample_datasets/redpajama_v2/mem_map/redpajama_v2_samples_1024_train.pbin
```
@rrutmann could you check if the tests run for you, if you change this to relative paths?

rrutmann replied:
Done in 595e007bea194fb4cad04b09502464c50f79222a

lllAlexanderlll commented on tests/data/checkpoint/checkpoint_for_testing/model_config.yaml:
```diff
@@ -0,0 +1,144 @@
+        dataset:
+          type_hint: PackedMemMapDatasetContinuous
+          config: 
+            raw_data_path: /home/richard-rutmann/s3/projects/remote_projects/modalities/data/sample_datasets/redpajama_v2/mem_map/redpajama_v2_samples_1024_test.pbin
```
@rrutmann here please too

rrutmann replied:
Done in 595e007bea194fb4cad04b09502464c50f79222a

lllAlexanderlll commented on tests/data/checkpoint/checkpoint_for_testing/model_config.yaml:
```diff
@@ -0,0 +1,144 @@
+  checkpointing_execution:
+    type_hint: FSDPToDiscCheckpointing
+    config: 
+      checkpoint_path: /home/richard-rutmann/s3/models/llm_gym/test/checkpoints
```
@rrutmann here as well

rrutmann replied:
Done in 595e007bea194fb4cad04b09502464c50f79222a

le1nux reviewed (changes requested):
Thanks for the PR 👍 Overall the implementation looks good. I left a couple of remarks and some suggestions.

rrutmann commented on src/modalities/__main__.py:
```diff
@@ -149,6 +154,31 @@ def entry_point_create_packed_data(src_path, dst_path, index_path, tokenizer_typ
     generator.run(dst_path)
 
 
+@main.command(name="convert_checkpoint")
```
Done in 84b74b9cbd060f2a80009f02bb78454a30bf5804

rrutmann commented on src/modalities/__main__.py:
```diff
@@ -149,6 +154,31 @@ def entry_point_create_packed_data(src_path, dst_path, index_path, tokenizer_typ
+    "--checkpoint_dir",
+    type=click_pathlib.Path(exists=False),
+    required=True,
+    help="Load checkpoint from this path.",
```
Done in 84b74b9cbd060f2a80009f02bb78454a30bf5804

rrutmann commented on src/modalities/__main__.py:
```diff
@@ -149,6 +154,31 @@ def entry_point_create_packed_data(src_path, dst_path, index_path, tokenizer_typ
+    "--output_path",
+    type=click_pathlib.Path(exists=False),
+    required=True,
+    help="Converted checkpoint will be written to this path.",
```
Done in 84b74b9cbd060f2a80009f02bb78454a30bf5804

rrutmann commented on src/modalities/__main__.py:
```diff
@@ -149,6 +154,31 @@ def entry_point_create_packed_data(src_path, dst_path, index_path, tokenizer_typ
+@main.command(name="convert_checkpoint")
+@click.option(
+    "--checkpoint_dir",
+    type=click_pathlib.Path(exists=False),
```
Done in 84b74b9cbd060f2a80009f02bb78454a30bf5804

rrutmann commented on src/modalities/config/config.py:
```diff
@@ -319,7 +320,7 @@ class AppConfig(BaseModel):
+class HuggingFaceModelConfig(PretrainedConfig):
     model_type = "modalities_gpt2"
 
     def __init__(self, config: GPT2Config = None, **kwargs):
```
We need a separate config class for every model architecture. E.g. model_type differs from model to model, but also the to_json_str-function. Would it be fine to rename HuggingFaceModelConfig to GPT2HuggingFaceModelConfig to be more clear?

le1nux replied:
Why do we need one Class per model architecture? Couldn't we pass in that information from outside?

rrutmann replied:
If I understand it correctly. you would like to pass in the to_json_str-function from outside? Doesn't this hurt the readability? And where should we then define this function?

rrutmann commented on src/modalities/models/gpt2/huggingface_model.py:
```diff
@@ -1,25 +1,25 @@
         # TODO offloading the parameters like this is ugly
-        self.model: GPT2LLM = GPT2LLM(**dict(config.config))
+        if model is None:
+            self.model: GPT2LLM = GPT2LLM(**dict(config.config))
```
Same as for the config: The forward- and prepare_inputs_for_generation-functions may differ for different model architectures. Maybe we rename HuggingFaceModel to GPT2HuggingFaceModel?

mrudat-iais replied:
@le1nux friendly reminder...

BlueCrescent replied:
This class and ModalitiesModelOutput should be at lest renamed as Richard proposed.
Ideally, there would be a way to at least have an abstract class that partially provides the Huggingface-Wrapping functionalities for our models. But this might be hard, especially since the parameters of forward() might change depending on the modalities of other models.

rrutmann replied:
We created an issue for that: 
https://github.com/Modalities/modalities/issues/71

rrutmann commented on tests/checkpointing/test_convert_to_hf_checkpoint.py:
```diff
@@ -0,0 +1,59 @@
+
+@pytest.fixture
+def config(config_path, monkeypatch):
+    monkeypatch.setenv("RANK", "0")
+    monkeypatch.setenv("LOCAL_RANK", "0")
+    monkeypatch.setenv("WORLD_SIZE", "1")
+    config_dict = load_app_config_dict(config_path)
```
These variables are required by pydantic, even in a non-distributed environment

rrutmann commented on tests/data/checkpoint/checkpoint_for_testing/optimizer.bin:
We don't need this for our tests, but future tests may require this. Do you suggest to remove it anyway?

le1nux replied:
I think we can remove it of now.

rrutmann replied:
Done in e97cb66aee6399695be9f0d302dc65404a67141a

le1nux commented on tests/data/checkpoint/checkpoint_for_testing/model.bin:
yes, that would be good :-) let's try to bring it down to less than 5MB.

mrudat-iais replied:
this is done now. we load the model, initialize randomly and only save the .bin file during the test, delete it afterwards.

rrutmann commented on src/modalities/__main__.py:
```diff
@@ -149,6 +154,31 @@ def entry_point_create_packed_data(src_path, dst_path, index_path, tokenizer_typ
+)
+def convert_checkpoint(checkpoint_dir, output_path):
+    checkpoint_dir = Path(checkpoint_dir)
+    config_file_path = checkpoint_dir / "model_config.yaml"
```
Done in fef53ffb33640c5c98a4bf16293db72e383aab8b

rrutmann commented on src/modalities/__main__.py:
```diff
@@ -163,6 +193,8 @@ def __init__(self, config: AppConfig) -> None:
         self.running_env: RunningEnv = self.resolvers.build_component_by_config(config=self.config.running_env)
 
+        self.model = None
+
```
Done in ce56c91ad6b6694c27133fccb7f000c0bc8f45c8

rrutmann commented on src/modalities/__main__.py:
```diff
@@ -351,6 +383,28 @@ def get_logging_publishers(
         return evaluation_result_publisher, batch_processed_publisher
 
+    def load_and_convert_checkpoint(self, checkpoint_path: Path, output_path: str):
+        self.model = self._get_model_from_checkpoint(checkpoint_path)
```
Done in ce56c91ad6b6694c27133fccb7f000c0bc8f45c8

rrutmann commented on src/modalities/__main__.py:
```diff
@@ -351,6 +383,28 @@ def get_logging_publishers(
+        model = checkpointing.load_model_checkpoint(model, checkpoint_path)
+        return model
+
+    def _convert_checkpoint(self, output_path, model):
```
Done in f4a4afedcf395e73a2b131330e521e003cd26aa2

rrutmann commented on src/modalities/checkpointing/checkpointing_execution.py:
```diff
@@ -203,3 +203,33 @@ def load_optimizer_checkpoint(self, optimizer: Optimizer, model: FSDP, file_path
+        raise NotImplementedError
+
+    def load_model_checkpoint(self, model: nn.Module, file_path: Path) -> nn.Module:
+        if self.global_rank == 0:
```
We intended to implement this in the same way it is implemented in FSDPToDiscCheckpointing

le1nux replied:
yes, `FSDPToDiscCheckpointing` requires a distributed environment. 
However, if we fix that checkpoint conversion never uses a torch dist environment then we can get rid of the rank assignments. This would make a lot of sense, because the checkpoints are stored as PyTorch checkpoints, so why the efforts of initialising the distributed environment?

rrutmann replied:
For the checkpoint conversion we don't need a distributed environment, that's true. And we don't initialize it, that was already implemented. However, the class PytorchToDiscCheckpointing could be used in contexts other than checkpoint conversion. And then it might lead to confusion, that the global rank check is not done in load_model_checkpoint, as it is for the class FSDPToDiscCheckpointing

rrutmann replied:
Done in 24698a005bd98713844f1ee772e142598f394ed2

rrutmann commented on src/modalities/__main__.py:
```diff
@@ -351,6 +383,28 @@ def get_logging_publishers(
+        if torch.distributed.is_initialized():
+            rank = torch.distributed.get_rank()
+        else:
+            rank = 0
```
Done in 3098431c560be5248408dd548ac08a1bafdc2b07

rrutmann commented on src/modalities/__main__.py:
```diff
@@ -351,6 +383,28 @@ def get_logging_publishers(
+        return model
+
+    def _convert_checkpoint(self, output_path, model):
+        config = HuggingFaceModelConfig(self.config.model.config)
```
Why should we add an instance variable as a parameter here? Also, how does it matter if we call self.config.model.config here or in load_and_convert_checkpoint?

rrutmann replied:
This is now also done with the latest changes

rrutmann commented on src/modalities/config/config.py:
```diff
@@ -319,7 +320,7 @@ class AppConfig(BaseModel):
+class HuggingFaceModelConfig(PretrainedConfig):
     model_type = "modalities_gpt2"
 
     def __init__(self, config: GPT2Config = None, **kwargs):
```
Done in 6464a7aadd045ffb7e35c2e4a6e3981580dc0b36

mrudat-iais commented:
update: additional merge conflicts are resolved.

lllAlexanderlll commented:
Locally all pre-commit hooks run through. However, the github action with pre-commit seems to fail for isort: ` Fixing /home/runner/work/modalities/modalities/src/modalities/logging_broker/subscriber_impl/results_subscriber.py`.

I uninstalled, installed, updated and cleaned the hooks locally, but the github action still fails. Any suggestions? @le1nux @luzian-hahn @spravil @mali-git @fromm-m @flxst

flxst commented:
> Locally all pre-commit hooks run through. However, the github action with pre-commit seems to fail for isort: ` Fixing /home/runner/work/modalities/modalities/src/modalities/logging_broker/subscriber_impl/results_subscriber.py`.
> 
> I uninstalled, installed, updated and cleaned the hooks locally, but the github action still fails. Any suggestions? @le1nux @luzian-hahn @spravil @mali-git @fromm-m @flxst

Have you tried to run `pre-commit run --all-files` manually? See #42.

lllAlexanderlll commented:
@flxst yes, I did, but all passes. I will look if the pre-commit hook versions differ between github actions and locally.

BlueCrescent commented on src/modalities/checkpointing/checkpoint_conversion.py:
```diff
@@ -0,0 +1,70 @@
+        self.checkpoint_path = checkpoint_dir / self.model_file_name
+
+    def convert_pytorch_to_hf_checkpoint(self):
+        # FIXME: make the conversion entry point configurable from outside:
+        # Which HuggingFaceAdapterConfig should be used etxactly currently it is too hard coded:
+        # Allow for a custom callable for conversion to HF to be given to the entrypoint
```
This should be addressed. This code should not depend at all on the GPT2 specific implementation. Or, if we can only write it in a model specific form. This whole file should be moved to the GPT2 directory.
Ideally, a more generic and if necessary, abstract implementation would be implemented.

lllAlexanderlll replied:
Will be addressed with issue #71

BlueCrescent commented on src/modalities/config/config.py:
```diff
@@ -246,6 +249,38 @@ class LLMDataLoaderConfig(BaseModel):
     skip_num_batches: Optional[int] = 0
 
 
+class HuggingFaceAdapterConfig(ABC, PretrainedConfig):
```
Consider moving the adapter configs into their own file. This file is already pretty long and the adapter configs are not strongly coupled with the other configs.

rrutmann replied:
We actually had these in different files, but were asked to merge them together with the other configs. So I would keep it here

BlueCrescent commented on src/modalities/checkpointing/checkpoint_conversion.py:
```diff
@@ -0,0 +1,70 @@
+
+    def _convert_checkpoint(self, model: torch.nn.Module, model_config: BaseModel):
+        hugging_face_model = HuggingFaceModel(config=model_config, model=model)
+        hugging_face_model.save_pretrained(self.output_hf_checkpoint_dir, safe_serialization=False)
```
Has black processed this file? I think it normally adds a new line at end of file.

rrutmann replied:
Actually there is a new line at the end of the file, I can see it locally in my IDE. For other files the final new line is also missing, so I think this is a displaying issue in Github

BlueCrescent commented on config_files/config.yaml:
```diff
@@ -120,8 +120,6 @@ checkpointing:
       k: -1 # -1 to save all checkpoints
   checkpointing_execution:
     type_hint: FSDPToDiscCheckpointing
-    config: 
-      checkpoint_path: /path/to/checkpoints
```
Does this need to be changed in the other config files as well?

BlueCrescent commented on src/modalities/__main__.py:
```diff
@@ -138,6 +142,41 @@ def entry_point_create_packed_data(src_path, dst_path, index_path, tokenizer_typ
     generator.run(dst_path)
 
 
+# FIXME
```
Does this FIXME still need to be addressed? What is it about?

rrutmann replied:
@lllAlexanderlll @mrudat-iais Do you know who put this line here? What should be fixed?

lllAlexanderlll commented on src/modalities/logging_broker/subscriber_impl/results_subscriber.py:
```diff
@@ -2,10 +2,10 @@
 from rich.console import Group
 from rich.panel import Panel
 
+import wandb
```
The new setup of the repo + linting resolved this.

lllAlexanderlll commented on src/modalities/checkpointing/checkpointing_execution.py:
```diff
@@ -216,3 +216,25 @@ def load_optimizer_checkpoint(self, optimizer: Optimizer, wrapped_model: FSDP, f
         return optimizer
+
+
+class PytorchToDiscCheckpointing(CheckpointingExecution):
```
As I understood @rrutmann this was done to be consistent in how to load a model checkpoint i.e. through `load_model_checkpoint` of a `CheckpointingExecution`, right?

rrutmann replied:
Excactly. We wanted to be in line with the general checkpoint API. And it might be that we add the missing functionalities later

lllAlexanderlll commented on src/modalities/checkpointing/checkpoint_conversion.py:
```diff
@@ -0,0 +1,70 @@
+
+        model = self._setup_model()
+        model = self._get_model_from_checkpoint(model)
+        model_config = GPT2HuggingFaceAdapterConfig(GPT2LLMConfig(**self.config_dict["model"]["config"]))
```
Done: https://github.com/Modalities/modalities/issues/81

lllAlexanderlll commented on src/modalities/config/config.py:
```diff
@@ -246,6 +249,38 @@ class LLMDataLoaderConfig(BaseModel):
+        raise NotImplementedError()
+
+
+class GPT2HuggingFaceAdapterConfig(HuggingFaceAdapterConfig):
```
Done

rrutmann commented on src/modalities/checkpointing/checkpoint_conversion.py:
```diff
@@ -0,0 +1,70 @@
+
+
+class CheckpointConversion:
+    def __init__(self, checkpoint_dir, config_file_name, model_file_name, output_hf_checkpoint_dir):
```
Done in 4953adf5bc89f749dd45eb053a03f45dccdb072f

rrutmann commented on src/modalities/__main__.py:
```diff
@@ -149,6 +154,31 @@ def entry_point_create_packed_data(src_path, dst_path, index_path, tokenizer_typ
+    required=True,
+    help="Converted checkpoint will be written to this path.",
+)
+def convert_checkpoint(checkpoint_dir, output_path):
```
Done in 3d5276ed330056896e701a1a4dd05f381f4553cf

rrutmann commented on src/modalities/__main__.py:
```diff
@@ -351,6 +404,27 @@ def get_logging_publishers(
+        model = checkpointing.load_model_checkpoint(model, checkpoint_path)
+        return model
+
+    def _convert_checkpoint(self, output_path: Union[str, Path], model: nn.Module):
```
Outdated by now, for other path variables we now only use Path

rrutmann commented on config_files/config.yaml:
Done in 3d5276ed330056896e701a1a4dd05f381f4553cf

rrutmann commented on src/modalities/checkpointing/checkpoint_conversion.py:
```diff
@@ -0,0 +1,70 @@
+        class ModelConfig(BaseModel):
+            model: PydanticModelIFType
+
+        self.components = self.component_factory.build_components(
```
Previously we needed this for our test, but not anymore. I changed it in 56151de1f564e08be48bc9903f54f0e839e39626

rrutmann commented on src/modalities/checkpointing/checkpoint_conversion.py:
```diff
@@ -0,0 +1,70 @@
+        model = checkpointing.load_model_checkpoint(model, self.checkpoint_path)
+        return model
+
+    def _convert_checkpoint(self, model: torch.nn.Module, model_config: BaseModel):
```
Done in 4bee4ff3b408582b7b84b5ec53f7885139f2b601

rrutmann commented on src/modalities/models/gpt2/huggingface_model.py:
```diff
@@ -0,0 +1,58 @@
+
+
+class HuggingFaceModel(PreTrainedModel):
+    config_class = GPT2HuggingFaceAdapterConfig
```
Yes, this is required and can't be removed

luzian-hahn commented on src/modalities/config/config.py:
```diff
@@ -246,6 +249,38 @@ class LLMDataLoaderConfig(BaseModel):
+    model_type = "modalities_gpt2"
+
+    def __init__(self, config: GPT2LLMConfig = None, **kwargs):
+        if isinstance(config, dict):
```
I was discussing this quickly with @lllAlexanderlll and the reason for not allowing a dict here is that it invites the user to bypass the pydantic validation in a dedicated, central initialization stage. 

Besides that [in this comment](https://github.com/Modalities/modalities/pull/34/files#diff-88a86f8a1ada3d5432fa5066602db0b26d2e2e6107bf48e6a24fc10c8db2cac0R40) there is the plan on dynamically initializing the config objects (and with those the actual functional objects). So passing it dict would not happen anyhow.

Finally you get advantages by limiting the datatype to a specific config object in your IDE, which allows e.g. easier automated refactorings like renamings and stuff.

luzian-hahn commented:
> https://github.com/Modalities/modalities/pull/34#discussion_r1478675749

I modified some type hints regarding this to prepare for the moment of abstracting the model init based on different hf-adapter types.
